src/attribution/onINP.ts

```diff
@@ -25,38 +28,224 @@ import {
+const entryToRenderTimeMap: WeakMap<
+  PerformanceEventTiming,
+  DOMHighResTimeStamp
+> = new WeakMap();
+
+// A reference to the idle task used to clean up entries from the above
+// variables. If the value is -1 it means no task is queue, and if it's
+// greater than -1 the value corresponds to the idle callback handle.
+let idleCallback: number = -1;
+
+/**
+ * Adds new LoAF entries to the `pendingLoAFs` list.
+ */
+const handleLoAFEntries = (entries: PerformanceLongAnimationFrameTiming[]) => {
+  entries.forEach((entry) => pendingLoAFs.push(entry));
+};
+
+/**
+ * Groups entries that were presented within the same animation frame by
+ * a common `renderTime`. This function works by referencing
+ * `pendingEntriesGroupMap` and using an existing render time if one is found
+ * (otherwise creating a new one). This function also adds all interaction
+ * entries to an `entryToRenderTimeMap` WeakMap so that the "grouped" render
+ * times can be looked up later.
+ */
+const groupEntriesByRenderTime = (entries: PerformanceEventTiming[]) => {
+  entries.forEach((entry) => {
+    let renderTime = entry.startTime + entry.duration;
+    let previousRenderTime;
+
+    // Iterate of all previous render times in reverse order to find a match.
+    // Go in reverse since the most likely match will be at the end.
+    for (let i = previousRenderTimes.length - 1; i >= 0; i--) {
+      previousRenderTime = previousRenderTimes[i];
+
+      // If a previous render time is within 8ms of the current render time,
+      // assume they were part of the same frame and re-use the previous time.
+      // Also break out of the loop because all subsequent times will be newer.
+      if (Math.abs(renderTime - previousRenderTime) <= 8) {
+        const group = pendingEntriesGroupMap.get(previousRenderTime)!;
+        group.startTime = Math.min(entry.startTime, group.startTime);
+        group.processingStart = Math.min(
+          entry.processingStart,
+          group.processingStart,
+        );
+        group.processingEnd = Math.max(
+          entry.processingEnd,
+          group.processingEnd,
+        );
+        group.entries.push(entry);
+
+        renderTime = previousRenderTime;
+        break;
+      }
+    }
+
+    // If there was no matching render time, assume this is a new frame.
+    if (renderTime !== previousRenderTime) {
+      previousRenderTimes.push(renderTime);
+      pendingEntriesGroupMap.set(renderTime, {
+        startTime: entry.startTime,
+        processingStart: entry.processingStart,
+        processingEnd: entry.processingEnd,
+        entries: [entry],
+      });
+    }
+
+    // Store the grouped render time for this entry for reference later.
+    if (entry.interactionId || entry.entryType === 'first-input') {
+      entryToRenderTimeMap.set(entry, renderTime);
+    }
+  });
+
+  // Queue cleanup of entries that are not part of any INP candidates.
+  if (idleCallback < 0) {
+    idleCallback = requestIdleCallback(cleanupEntries);
```

Comment:
Do we need a fallback if `requestIdleCallback` is not supported?

I mean that shouldn't happen as all browsers that support INP support that and Safari has added it, but still feels like a guard, and fallback to immediate call, could help here.

Reply:
> Do we need a fallback if `requestIdleCallback` is not supported?

I don't think so? Both Firefox and Chrome support this API, so the only risk would be if Safari suddenly adds support for Event Timing but doesn't enable rIC for some reason. And at that point we could just update the library to "enable Safari support". WDYT?

Reply:
Ok, I added (because the fallback is trivial), but I think we should revisit this if/when Safari ships support.

As a meta point, I don't want this library (or any public code our team creates) to promote the sense that modern web APIs are not 100% reliable and that you should be weary of using them without fallbacks, transpiling, bundling polyfills, etc.